substrate/primitives/io/src/lib.rs

```diff
@@ -2712,55 +2711,289 @@ pub trait Crypto {
+	/// keystore.
+	fn ecdsa_bls381_num_public_keys(&mut self, id: PassPointerAndReadCopy<KeyTypeId, 4>) -> u32 {
+		self.extension::<KeystoreExt>()
+			.expect("No `keystore` associated for the current context!")
+			.ecdsa_bls381_public_keys(id)
+			.len() as u32
+	}
+
+	/// Returns the `ecdsa & bls12-381` public key for the given key type and index in the keystore.
+	/// Panics if the key index is out of bounds.
+	fn ecdsa_bls381_public_key(
+		&mut self,
+		id: PassPointerAndReadCopy<KeyTypeId, 4>,
+		index: u32,
+		out: PassPointerAndWrite<&mut ecdsa_bls381::Public, 177>,
+	) {
+		out.0.copy_from_slice(
+			self.extension::<KeystoreExt>()
+				.expect("No `keystore` associated for the current context!")
+				.ecdsa_bls381_public_keys(id)
+				.get(index as usize)
+				.expect("Key index out of bounds!"),
+		);
+	}
+
+	/// A convenience wrapper providing a developer-friendly interface for the obsoleted
+	/// `ecdsa_bls381_public_keys` host function
+	#[wrapper]
+	fn ecdsa_bls381_public_keys(id: KeyTypeId) -> Vec<ecdsa_bls381::Public> {
+		let num_keys = ecdsa_bls381_num_public_keys(id);
+		let mut keys = Vec::new();
+		for i in 0..num_keys {
+			let mut key = ecdsa_bls381::Public::default();
+			ecdsa_bls381_public_key(id, i, &mut key);
+			keys.push(key);
+		}
+		keys
+	}
+
+	/// Sign the given `msg` with the `ecdsa & bls12-381` key that corresponds to the given public
+	/// key and key type in the keystore.
+	///
+	/// Stores the signature in the provided output buffer.
+	/// Returns 0 on success, -1 on error.
+	#[wrapped]
+	fn ecdsa_bls381_sign(
+		&mut self,
+		id: PassPointerAndReadCopy<KeyTypeId, 4>,
+		pub_key: PassPointerAndRead<&ecdsa_bls381::Public, 177>,
+		msg: PassFatPointerAndRead<&[u8]>,
+		out: PassPointerAndWrite<&mut ecdsa_bls381::Signature, 177>,
+	) -> ConvertAndReturnAs<Result<(), ()>, RIIntResult<VoidResult, VoidError>, i64> {
 		self.extension::<KeystoreExt>()
 			.expect("No `keystore` associated for the current context!")
-			.ecdsa_bls381_generate_new(id, seed)
-			.expect("`ecdsa_bls381_generate` failed")
+			.ecdsa_bls381_sign(id, pub_key, msg)
+			.ok()
+			.flatten()
+			.map(|sig| {
+				out.0.copy_from_slice(&sig);
+			})
+			.ok_or(())
+	}
+
+	/// A convenience wrapper providing a developer-friendly interface for the `ecdsa_bls381_sign`
+	/// host function.
+	#[wrapper]
+	fn ecdsa_bls381_sign(
+		id: KeyTypeId,
+		pub_key: &ecdsa_bls381::Public,
+		message: &[u8],
+	) -> Option<ecdsa_bls381::Signature> {
+		let mut signature = ecdsa_bls381::Signature::default();
+		ecdsa_bls381_sign__wrapped(id, pub_key, message, &mut signature).ok()?;
+		Some(signature)
+	}
+
+	/// Sign the given `msg` with the `ecdsa & bls12-381` key that corresponds to the given public
+	/// key and key type in the keystore using keccak256 hash for the ECDSA component.
+	///
+	/// Stores the signature in the provided output buffer.
+	/// Returns 0 on success, -1 on error.
+	#[wrapped]
+	fn ecdsa_bls381_sign_with_keccak256(
```

Comment:
same here, bls doesn't sign with sha3 it is only ecdsa so we can simulate the action with other functions.